src/Sarif/Visitors/RebaseUriVisitor.cs

```diff
@@ -87,8 +93,37 @@ public override Run VisitRun(Run node)
         {
             _files = node.Files;
 
+            // Save the file key in the property bag so we can remap it during rebasing.
+            foreach (string key in _files.Keys)
+            {
+                _files[key].SetProperty(OriginalFileKey, key);
+            }
+
             Run newRun = base.VisitRun(node);
 
+            // Remove the remapped files (they are duplicated during remapping).
+            foreach (string key in _files.Keys.ToArray())
+            {
+                string originalFileKey = _files[key].GetProperty<string>(OriginalFileKey);
+
+                if (originalFileKey != null)
+                {
```

Comment:
>{ [](start = 16, length = 1)

I see what's going on here. Both the original entry and the duplicate have the `OriginalFileKey` property. Get rid of the original entry altogether, and as for the duplicate (the new one), get rid of the property.

So two things:
1. This is worth a comment.
2. BUG: If you apply this transformation to a file that has _already_ been remapped (that is, its `run.files` dictionary keys are already of the form `#<base>#<relativeUri>`). In that case, you will _not_ actually duplicate the dictionary entry; when you get here you'll find that there is exactly one entry, and it is the original entry, and you're going to delete it.

Reply:
(You'll definitely need a test case for <span>#</span>2.)

---
In reply to: [218974200](https://github.com/Microsoft/sarif-sdk/pull/1098#discussion_r218974200) [](ancestors = 218974200)

Reply:
You could avoid this, I think. Why are you leaving the duplicate entries to be cleaned up later? At the point you create the duplicate (L165), you already know that you don't need the old node -- and you don't need the `OriginalFileKey` property on the new node any more either. Get rid of them both, right then, and you won't need the post-processing in `VisitRun` (the whole code block starting on L105). Then you're in no danger of clobbering nodes that weren't duplicated.

---
In reply to: [218974651](https://github.com/Microsoft/sarif-sdk/pull/1098#discussion_r218974651) [](ancestors = 218974651,218974200)

Reply:
Oh crud. You can't do that while you're in the process of enumerating the keys. Here's what you can do.

1. As soon as you create the new node:
    1. Remove the `OriginalFileKey` property from the new node.
    2. Mark the old node with a property `ToBeRemoved`.
2. In the `VisitRun` post-processing:
    - Remove any node with the `ToBeRemoved` property.

---
In reply to: [218978128](https://github.com/Microsoft/sarif-sdk/pull/1098#discussion_r218978128) [](ancestors = 218978128,218974651,218974200)

Reply:
Hah hah. You don't need the `ToBeRemoved` property. Instead:

1. As soon as you create the new node, remove its `OriginalFileKey` property.
2. In the `VisitRun` post-processing, remove any node that still has the `OriginalFileKey` property. (Those are the old nodes.)

---
In reply to: [218981486](https://github.com/Microsoft/sarif-sdk/pull/1098#discussion_r218981486) [](ancestors = 218981486,218978128,218974651,218974200)